common/map_store.c

```diff
@@ -0,0 +1,504 @@
+#include "config.h"
+#include <assert.h>
+#include <ccan/bitops/bitops.h>
+#include <ccan/crypto/siphash24/siphash24.h>
+#include <ccan/endian/endian.h>
+#include <ccan/err/err.h>
+#include <ccan/htable/htable_type.h>
+#include <ccan/mem/mem.h>
+#include <ccan/noerr/noerr.h>
+#include <common/gossip_store.h>
+#include <common/map_store.h>
+#include <common/node_id.h>
+#include <common/pseudorand.h>
+#include <common/type_to_string.h>
+#include <common/utils.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <wire/gen_peer_wire.h>
+
+static const struct short_channel_id *chan_id(const struct mapped_chan *c)
+{
+	return &c->scid;
+}
+static bool chan_eq_id(const struct mapped_chan *c,
+		       const struct short_channel_id *scid)
+{
+	return short_channel_id_eq(&c->scid, scid);
+}
+static size_t scid_hash(const struct short_channel_id *scid)
+{
+	return siphash24(siphash_seed(), scid, sizeof(*scid));
+}
+HTABLE_DEFINE_TYPE(struct mapped_chan, chan_id, scid_hash, chan_eq_id,
+		   chan_htable);
+
+struct gossip_map {
+	/* The file descriptor and filename to monitor */
+	int fd;
+	int st_dev, st_ino;
+	const char *fname;
+
+	/* The memory map of the file: u8 for arithmetic portability */
+	u8 *map;
+	/* map_end is where we read to so far, map_size is total size */
+	size_t map_end, map_size;
+
+	/* Map of node id -> node */
+	struct htable nodes;
+
+	/* Map of short_channel_id id -> channel */
+	struct chan_htable channels;
+
+	/* Array of nodes, so we can use simple index. */
+	struct mapped_node *node_arr;
+
+	/* Array of chans, so we can use simple index */
+	struct mapped_chan *chan_arr;
```

Comment:
Hmmm.  So if I have two plugins accessing the routemap, they will have separate arrays?  Will this not increase the memory footprint?

Reply:
I was thinking they might want completely different representations. The current one are designed for routing as we have now, but even then it needs to be extended.

Plus the overhead is only 10.5 mb, which is pretty good.